crates/common/src/raindex_client/vaults_list.rs

```diff
@@ -0,0 +1,424 @@
+use alloy::{primitives::Bytes, sol_types::SolCall};
+use rain_math_float::Float;
+use rain_orderbook_bindings::OrderBook::multicallCall;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use wasm_bindgen_utils::prelude::*;
+
+use crate::raindex_client::vaults::RaindexVault;
+use once_cell::sync::Lazy;
+
+static ZERO_FLOAT: Lazy<Float> = Lazy::new(|| Float::parse("0".to_string()).unwrap());
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[wasm_bindgen]
+pub struct VaultsList(Vec<RaindexVault>);
+
+impl VaultsList {
+    pub fn new(vaults: Vec<RaindexVault>) -> Self {
+        Self(vaults)
+    }
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+    pub fn get_vaults(&self) -> &Vec<RaindexVault> {
+        &self.0
+    }
+
+    pub fn get_withdrawable_vaults(&self) -> Vec<&RaindexVault> {
+        self.0
+            .iter()
+            .filter(|vault| vault.balance().gt(*ZERO_FLOAT).unwrap_or(false))
+            .collect()
+    }
+    pub async fn get_withdraw_calldata(&self) -> Result<Bytes, VaultsListError> {
+        let mut calldatas: Vec<Bytes> = Vec::new();
+        let vaults_to_withdraw = self.get_withdrawable_vaults();
+        // If no vaults to withdraw, return error
+        if vaults_to_withdraw.is_empty() {
+            return Err(VaultsListError::NoWithdrawableVaults);
+        }
+
+        // Generate multicall calldata for all vaults
+        for vault in vaults_to_withdraw {
+            let amount = vault.balance().format().map_err(|e| {
+                VaultsListError::WithdrawMulticallError(format!("Failed to format balance: {}", e))
+            })?;
+            match vault.get_withdraw_calldata(amount.clone()).await {
+                Ok(calldata) => calldatas.push(calldata),
+                Err(e) => return Err(VaultsListError::WithdrawMulticallError(e.to_readable_msg())),
+            }
+        }
+
+        if calldatas.len() == 1 {
+            // If only one vault, return its calldata directly
+            return Ok(calldatas[0].clone());
+        }
+        // Otherwise, create a multicall with all vaults' calldata
+        let multicall = multicallCall { data: calldatas };
+        let encoded = multicall.abi_encode();
+        Ok(encoded.into())
+    }
+}
+
+#[wasm_export]
+impl VaultsList {
+    /// Creates a new VaultsList from an array of RaindexVault objects
+    ///
+    /// ## Examples
+    ///
+    /// ```javascript
+    /// const vaultsList = VaultsList.new([vault1, vault2, vault3]);
+    /// ```
+    #[wasm_export(
+        js_name = "new",
+        return_description = "A new VaultsList instance",
+        preserve_js_class
+    )]
+    pub fn new_wasm(
```

Comment:
this is not used in the applications so i think it can be removed

Reply:
OK